cmd/car/extract.go

```diff
@@ -139,99 +168,144 @@ func resolvePath(root, pth string) (string, error) {
 	return joined, nil
 }
 
-func extractDir(c *cli.Context, ls *ipld.LinkSystem, n ipld.Node, outputRoot, outputPath string) error {
-	dirPath, err := resolvePath(outputRoot, outputPath)
-	if err != nil {
-		return err
+func extractDir(c *cli.Context, ls *ipld.LinkSystem, n ipld.Node, outputRoot, outputPath string, matchPath []string) (int, error) {
+	if outputRoot != "" {
+		dirPath, err := resolvePath(outputRoot, outputPath)
+		if err != nil {
+			return 0, err
+		}
+		// make the directory.
+		if err := os.MkdirAll(dirPath, 0755); err != nil {
+			return 0, err
+		}
 	}
-	// make the directory.
-	if err := os.MkdirAll(dirPath, 0755); err != nil {
-		return err
+
+	if n.Kind() != ipld.Kind_Map {
+		return 0, ErrNotDir
 	}
 
-	if n.Kind() == ipld.Kind_Map {
-		mi := n.MapIterator()
-		for !mi.Done() {
-			key, val, err := mi.Next()
-			if err != nil {
-				return err
-			}
-			ks, err := key.AsString()
-			if err != nil {
-				return err
-			}
-			nextRes, err := resolvePath(outputRoot, path.Join(outputPath, ks))
+	subPath := matchPath
+	if len(matchPath) > 0 {
+		subPath = matchPath[1:]
+	}
+
+	extractElement := func(name string, n ipld.Node) (int, error) {
+		var nextRes string
+		if outputRoot != "" {
+			var err error
+			nextRes, err = resolvePath(outputRoot, path.Join(outputPath, name))
 			if err != nil {
-				return err
+				return 0, err
 			}
 			if c.IsSet("verbose") {
 				fmt.Fprintf(c.App.Writer, "%s\n", nextRes)
 			}
+		}
 
-			if val.Kind() != ipld.Kind_Link {
-				return fmt.Errorf("unexpected map value for %s at %s", ks, outputPath)
-			}
-			// a directory may be represented as a map of name:<link> if unixADL is applied
-			vl, err := val.AsLink()
-			if err != nil {
-				return err
-			}
-			dest, err := ls.Load(ipld.LinkContext{}, vl, basicnode.Prototype.Any)
-			if err != nil {
-				return err
+		if n.Kind() != ipld.Kind_Link {
+			return 0, fmt.Errorf("unexpected map value for %s at %s", name, outputPath)
+		}
+		// a directory may be represented as a map of name:<link> if unixADL is applied
+		vl, err := n.AsLink()
+		if err != nil {
+			return 0, err
+		}
+		dest, err := ls.Load(ipld.LinkContext{}, vl, basicnode.Prototype.Any)
+		if err != nil {
+			if nf, ok := err.(interface{ NotFound() bool }); ok && nf.NotFound() {
+				fmt.Fprintf(c.App.ErrWriter, "data for directory entry not found: %s (skipping...)\n", name)
+				return 0, nil
```

Comment:
Highlighting this while I think about it and so it doesn't get lost in the code move—instead of erroring where a block load fails, it skips whatever it's failed to load. This needs more testing, and perhaps it's not feasable, because I've got an example where this partially works but then fails with an iterator overrun error which I don't yet understand. I'm guessing it's to do with sharded directory incompleteness but I need to work that out.